storage/netapp_e/na_eseries_facts.py

```diff
@@ -0,0 +1,194 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+import json
+
+from ansible.module_utils.api import basic_auth_argument_spec
+from ansible.module_utils.basic import AnsibleModule, get_exception
+from ansible.module_utils.urls import open_url
+
+try:
+    from urllib.error import HTTPError
+except ImportError:
+    from urllib2 import HTTPError
+
+DOCUMENTATION = '''
+module: na_eseries_facts
+version_added: '2.2'
+short_description: Get facts about NetApp E-Series arrays
+extends_documentation_fragment:
+        - auth_basic
+description:
+    - Return various information about NetApp E-Series storage arrays (eg, configuration, disks)
+
+author: Kevin Hulquest (@hulquest)
+'''
+
+EXAMPLES = """
+---
+    - name: Get array facts
+      na_eseries_facts:
+        array_id: "{{ netapp_array_id }}"
+        api_url: "{{ netapp_api_url }}"
+        api_username: "{{ netapp_api_username }}"
+        api_password: "{{ netapp_api_password }}"
+        validate_certs: "{{ netapp_api_validate_certs }}"
+"""
+
+RETURN = """
+msg: Gathered facts for <StorageArrayId>.
+"""
+
+
+def request(url, data=None, headers=None, method='GET', use_proxy=True,
+            force=False, last_mod_time=None, timeout=10, validate_certs=True,
+            url_username=None, url_password=None, http_agent=None, force_basic_auth=True, ignore_errors=False):
+    try:
+        r = open_url(url=url, data=data, headers=headers, method=method, use_proxy=use_proxy,
+                     force=force, last_mod_time=last_mod_time, timeout=timeout, validate_certs=validate_certs,
+                     url_username=url_username, url_password=url_password, http_agent=http_agent,
+                     force_basic_auth=force_basic_auth)
+    except HTTPError:
+        err = get_exception()
+        r = err.fp
+
+    try:
+        raw_data = r.read()
+        if raw_data:
+            data = json.loads(raw_data)
+        else:
+            data = None
+    except:
+        if ignore_errors:
+            pass
+        else:
+            raise
+
+    resp_code = r.getcode()
+
+    if resp_code >= 400 and not ignore_errors:
+        raise Exception(resp_code, data)
+    else:
+        return resp_code, data
+
+
+def main():
+    argument_spec = basic_auth_argument_spec()
+    argument_spec.update(
+        ssid=dict(required=True))
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True
+    )
+
+    p = module.params
+
+    ssid = p['ssid']
+    validate_certs = p['validate_certs']
+
+    try:
+        api_usr = p['api_username']
+        api_pwd = p['api_password']
+        api_url = p['api_url']
+    except KeyError:
+        import os
+        api_usr = os.environ['API_USERNAME']
+        api_pwd = os.environ['API_PASSWORD']
+        api_url = os.environ['API_URL']
```

Comment:
Cool.  I can see the utility of falling back to the environment.  We have something similar for AWS and openstack moudles I think.  Those are done in module_utils/ libraries for that type of module though.  As I'm thinking about this more, if you want to have environment variables in the future, they probably would have to use names that reference netapp.  One module might use basic auth for netapp and another for a web based notification service.  The user probably wouldn't have the same username, password, and URL for both.